### PR TITLE
Implementation/45339 modify the backend to support saving the non working days

### PR DESCRIPTION
**Pull request by dombesz**
https://community.openproject.org/work_packages/45339

Todo:
- [x] Create `::Settings::WorkingDayUpdateService` and batch update the NonWorkingDays
- [x] Adapt the `WorkingDaysSettingsController` to use the `::Settings::WorkingDaysUpdateService`.
- [x] Bubble up the `modified_non_working_days` array when the form saving is failed
- [x] ~~Update the `WorkPackages::ApplyWorkingDaysChangeJob` to take `NonWorkingDays` into account.~~ See [45473](https://community.openproject.org/work_packages/45473)

### CODE REVIEW

**Comment by ulferts** [comment]
The `dev` branch had failing specs from whence you branched off which is probably why the specs are failing in this PR. Please rebase/merge to fix this.

File: spec/services/settings/working_days_update_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    shared_examples 'unsuccessful working days settings call' do
+      include_examples 'unsuccessful call'
+
+      context 'when non working days are present' do
+        include_examples 'unsuccessful call'
+
+        it 'does not persists the non working days' do
+          expect { subject }.not_to change(NonWorkingDay, :count)
+        end
+      end
```
Comment (cbliard) [inline spec/services/settings/working_days_update_service_spec.rb]: I'm not sure this nested context is needed. It looks like the shared examples `'unsuccessful call'` are called twice. Maybe the `it 'does not persist the non-working days'` should be moved one level up, and the `context 'when non working days are present'` removed.

**Review by cbliard — CHANGES_REQUESTED** [review]
I ran the test `app/services/settings/working_days_update_service.rb` on my machine on 2022-12-27, and by luck I got a failing test where:
- New Year Eve non working day was created on 2022-12-31
- The non working day to delete was also 2022-12-31

This ended up in failing to create the non-working day because a record with that date already existed in the database (db constraint).

The case of deleting a non-working day and creating one having the same day as the one being deleted should be covered by a test. I see two possible implementations:
- Either the contract and denies doing so with a proper error message.
- Or the deletions happen first, so that the creation works afterwards.

The desired behavior depends on how the specification is interpreted.

Apart from that, it looks like it works as expected. It is hard to tell until there is a GUI to try it out.

File: spec/services/settings/working_days_update_service_spec.rb (hunk 1 above)
Comment (dombesz) [reply spec/services/settings/working_days_update_service_spec.rb]: Removed.

**Comment by dombesz** [comment]
@cbliard thanks for the review, and catching the edge case of deleting and re-creating a non working day, I addressed it here 526f61c21eb7d6aa559fcd1d40362fcec63fa6c2 . Yes it is a bit harder without the UI, but the UI PR is on the way and I expect to handle more unexpected cases in that PR.

**Review by cbliard — APPROVED** [review]

**Pull request merged**